libnmstate/nm/checkpoint.py

```diff
@@ -149,31 +116,145 @@ def _refresh_checkpoint_timeout(self, _user_data):
+                self._remove_checkpoint_refresh_timeout()
+            logging.debug(f"Checkpoint {self._dbuspath} rollback executed")
+            self._ctx.wait_all_finish()
 
     @property
     def dbuspath(self):
         return self._dbuspath
+
+    def _checkpoint_create_callback(self, client, result, data):
+        try:
+            cp = client.checkpoint_create_finish(result)
+            self._dbuspath = cp.get_path()
+            if cp:
+                logging.debug(
+                    "Checkpoint {} created for all devices".format(
+                        self._dbuspath
+                    )
+                )
+                self._ctx.finish_async("Create checkpoint")
+            else:
+                error_msg = (
+                    f"dbuspath={self._dbuspath} "
+                    f"timeout={self._timeout} "
+                    f"callback result={cp}"
+                )
+                self._ctx.fail(
+                    NmstateLibnmError(f"Checkpoint create failed: {error_msg}")
+                )
+        except Exception as e:
+            self._ctx.fail(
+                NmstateLibnmError(f"Checkpoint create failed: error={e}")
+            )
+
+    def _checkpoint_rollback_callback(self, client, result, data):
+        action = data
+        try:
+            self._check_rollback_result(client, result, self._dbuspath)
+            self._dbuspath = None
+            self._ctx.finish_async(action)
+        except Exception as e:
+            self._ctx.fail(
+                NmstateLibnmError(f"Checkpoint rollback failed: error={e}")
+            )
+
+    def _check_rollback_result(self, client, result, dbus_path):
+        ret = client.checkpoint_rollback_finish(result)
+        logging.debug(f"Checkpoint {dbus_path} rollback executed")
+        for path in ret:
+            nm_dev = client.get_device_by_path(path)
+            iface = path if nm_dev is None else nm_dev.get_iface()
+            if (
+                nm_dev.get_state() == common.NM.DeviceState.DEACTIVATING
+                and nm_dev.get_state_reason()
+                == common.NM.DeviceStateReason.NEW_ACTIVATION
+            ):
+                action = f"Waiting for rolling back {iface}"
+                self._ctx.register_async(action)
+                handler_id = nm_dev.connect(
+                    NM_DEV_STATE_CHANGED_SIGNAL,
+                    self._wait_nm_dev_ready_callback,
+                    action,
+                )
+                self._handlers[nm_dev.get_iface()] = handler_id
+            if ret[path] != 0:
+                logging.error(f"Interface {iface} rollback failed")
+            else:
+                logging.debug(f"Interface {iface} rollback succeeded")
+
+    def _wait_nm_dev_ready_callback(
+        self, nm_dev, state, last_state, reason, action
+    ):
+        if self._ctx.is_cancelled():
+            print("CANCELED!!")
+            return
+        if state == common.NM.DeviceState.ACTIVATED:
```

Comment:
This is incorrect.
We should follow the same rule on activation: don't wait on master interface, don't wait DHCP, etc.

Please find a way to unify this action with code in nm/active_connection.py and nm/connection.py.

Reply:
I agree. Look at the new approach. Thanks!